### ISSUE TITLE
Add allocation-free iterators

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/roaring#94]
Currently, iterators are implemented by instantiating sub-iterators that run over the containers. This approach requires us to create many more temporary objects than necessary.

A much simpler approach is the one taken by the CRoaring library where no memory allocation is done...
https://github.com/RoaringBitmap/CRoaring/blob/master/src/roaring.c#L1156

This should reduce the amount of code and reduce the pressure on the garbage collector, which should improve performance in real-life systems.

**Comment by steveyen** [comment]
Hi roaring team,

I've run into this issue of GC pressure, too, in a roaring-based backend for the github.com/blevesearch/bleve full-text library.  The roaring library's already given some great perf gains, and we're looking for even more.  Reducing GC pressure has been a recurring perf dragon to slay.

Each term in a bleve inverted index is represented by a separate roaring bitmap, and during searches, it roughly ends up doing something like this...

     allocate a new roaringBitmap; or reuse a previous one (hopefully) via roaringBitmap.Clear()

     roaringBitmap.FromBuffer(someMemoryMappedData)

     optionally perform some roaring And() / Or()'s depending on the search criteria

     iterator := roaringBitmap.Iterator()

     walk thru the iterator

     repeat the above, trying to reuse the roaringBitmap for whatever next term needs to be searched

I've tried some speculative, experimental changes to roaring to try to make that better, here at  https://github.com/steveyen/roaring/commits/optimize-bitmap-fromBuffer-iterator-reuse

As a summary handwave, those experimental changes try to improve roaring's Clear() and FromBuffer() implementations to reuse more internal data structures.

And, I experimented with adding a speculative roaringBitmap.IteratorReuse() method, so that bleve can do the following...

     iterator = roaringBitmap.IteratorReuse(iterator)

     walk thru the iterator

Those kinds of perf improvements bring some bleve micro-benchmarking from 163K queries/sec to 180K queries/sec after the changes.

But, I imagine there might be some concerns about this, including...

A) I sense that I probably got the copyOnWrite semantics wrong during the Clear() / FromBuffer() changes, as copyOnWrite feels subtle.

B) The proposed new method named IteratorReuse() was the quick name I came up with on short notice, and might not meet your tastes about this or how memory reuse ought to work.  Perhaps, for example, this all can be done without any API changes?

C) I took a peek at the CRoaring code linked above, but now realize it's been too many years since I've read c code.

D) Whatever else I'm missing here. :-)

Would be happy to help make improvements here happen, following whatever guidance / feedback y'all might have.

cheers,
steve

**Comment by steveyen** [comment]
Adding a link to issue https://github.com/RoaringBitmap/roaring/issues/125, as there's discussion there of also perhaps improving the Clear() implementation

**Comment by lemire** [comment]
> Perhaps, for example, this all can be done without any API changes?

I expect that there is no harm in additive changes to the API. Thus we could create a new iterator approach.

The Java version of Roaring has ["flyweight" iterators](https://github.com/RoaringBitmap/RoaringBitmap/blob/master/roaringbitmap/src/main/java/org/roaringbitmap/IntIteratorFlyweight.java). I expect it does pretty much what you are doing.  Have you had a look at it?

The term flyweight seems to be the correct technical term:

https://en.wikipedia.org/wiki/Flyweight_pattern

I realize that you might have done several optimizations... but we need to start somewhere... it feels like iterators and GC pressure is a good starting point.

Would you consider doing a PR regarding these iterators?

**Comment by steveyen** [comment]
> Would you consider doing a PR regarding these iterators?

Sure, would be happy to get something going here.  And, thanks for the link to the java flyweight iterators -- following that pattern would also work.

Before I run that down, though, double-checking... the quick & dirty approach I had taken was by adding new methods of the same name to each of the various container/iterator implementations.

The flyweight approach, in contrast, seems to have less "encapsulation", where the flyweight seems to "know" a little bit about the internals of each backend container implementation.

Figured I should double check that that's your preference before running down the flyweight way.

**Comment by lemire** [comment]
@steveyen 

Yes, the approach used in Java (flyweight) requires tighter integration but I submit to you that it does a better job at achieving your objectives. I also submit to you that tight integration is not harmful in this instance.

The idea is that you create a flyweight iterator and once you have done so, no memory allocation is needed because you have one instance of each of the three possible short iterators.

In your approach, if I read it right, you opportunistically attempt to reuse iterators, but you will still trigger garbage collection.

Do you agree with my assessment?

**Comment by maciej** [comment]
Are allocation-heavy iterators still a problem now that we have `ManyIterator`?

**Comment by lemire** [comment]
@maciej I think it is an orthogonal issue.

Even with ``ManyIterator``, you need to create, and eventually garbage-collect, some kind of data structure. And this data structure generates other data structures like ``bitmapContainerManyIterator``. This does not cause too much trouble in microbenchmarks, but in real systems, you can end up spending a lot of time allocating and collecting iterators. It is not *always* a problem... but when it is, you'd like to have a way out.

The way out is simple: construct some data structure that keeps reusing the same instances (no more memory allocation). It is not as nice from a software engineering point of view, it makes your code more difficult to write... (e.g., you don't want to reuse the same iterator object over many threads!!!), but it could be worth it for the speed.

What ``ManyIterator`` does that is beneficial, I think, is to reduce the number of function calls. And that's especially beneficial in Go.

**Comment by Oppen** [comment]
Alternatively, the container interface can have a `Next(current uint16) int32` method, so the iterator would implement `Next() uint16` somewhat like this (names of types weren't checked, it's just the idea):
```go
func (it *iterator) Next() {
    next := it.currentContainer.Next(it.currentValue)
    if next == -1 { // Flag abscense
        it.updateContainer() // For simplicity.
        return it.Next()
    }
    it.currentValue = next
    return next
}
```
You get the benefits of using interfaces to separate logic without extra allocations.

**Comment by Oppen** [comment]
Typo: pass the low bits only.

**Comment by Oppen** [comment]
We could also store the index inside the container with a two-value return and pass the index directly, but in any case a gallop/shotgun search for the value would probably be good enough for arrays and runs and the value and the index are the same for bitsets.

**Comment by lemire** [comment]
@Oppen 

Let us consider...

```
next := it.currentContainer.Next(it.currentValue)
```

Ok. So if you have an array container, how would `it.currentContainer.Next(it.currentValue)` work? Probably you would start with a binary search. That's not great. So for performance you want to maintain more of a state.

**Comment by lemire** [comment]
> a gallop/shotgun search for the value would probably be good enough for arrays and runs and the value and the index are the same for bitsets.

I am skeptical. I expect it would cause a significant performance penalty.

**Comment by Oppen** [comment]
Maybe. Worst case you need to pass an index in addition to a value (runs would lose information without the value), and keep that index as the only extra data in the iterator. I think that's abstract enough, considering any kind of container will need a concept of numbered element anyway.

**Comment by Oppen** [comment]
The signature would instead be something like `Next(fromIndex, fromValue uint16_t) int32`.

**Comment by Oppen** [comment]
I know, but then we'd still need to get the types. In C they're an explicit part of the `roaring_array_t` structure, but in Go we use interfaces I think. So, either we do type switches needlessly or we implement typed functions anyway to match that implementation. I think it makes sense to just use an interface the containers can implement (but with no intermediate iterator).
The higher level iterator would still contain the necessary extra data for invocation.

**Comment by Oppen** [comment]
I can send a PR tomorrow to illustrate, and run the benchmarks to see if it regresses. I won't be able to see after work today because I'll have an eyes background (I'm not sure if this is the proper translation to English, but I hope it's clear enough).

**Comment by jacksonrnewhouse** [comment]
I think the concerns about allocating structs for iterators are likely somewhat overblown. If you're using them to perform some calculation across all elements it *should* not escape and be allocated on the stack, which is very cheap. If you're embedding it in some other struct to call sporadically, it'll be on the heap, but given that you've allocated that struct likely fine. Let me know if I'm missing something

**Comment by Oppen** [comment]
Iterators do escape:
```
./arraycontainer.go:39:9: &shortIterator{...} escapes to heap
./arraycontainer.go:42:7: leaking param content: ac
./arraycontainer.go:43:9: &reverseIterator{...} escapes to heap
./arraycontainer.go:46:7: leaking param content: ac
./arraycontainer.go:47:9: &shortIterator{...} escapes to heap
```

**Comment by lemire** [comment]
@Oppen Sadly, pretty much everything escapes in Go. 

:-(

**Comment by Oppen** [comment]
I know, but I wanted to provide data for my claim :)

**Comment by jacksonrnewhouse** [comment]
I'm on vacation so don't have access to run my own escape analyses. But, given that these are private methods allowing escape, I think we could do better.

Part of my confidence comes from my experimentation in roaring bitmaps mainly backed by a single byte slice. It instantiates local containers had

**Comment by jacksonrnewhouse** [comment]
Sorry, I was typing a long response on my phone and fat fingered

**Comment by jacksonrnewhouse** [comment]
The summary of my response is that, given these are internal methods, I think we may be able to change them so they are stack allocated

**Comment by lemire** [comment]
>  I think we may be able to change them so they are stack allocated

I think that this would almost sure be best.

**Comment by Oppen** [comment]
A less disruptive alternative could be using `sync.Pool`s for the containers' iterators. We would then only add a release method to the interface and call it before changing iterators.

**Comment by lemire** [comment]
> A less disruptive alternative could be using sync.Pools for the containers' iterators. We would then only add a release method to the interface and call it before changing iterators.

It is easy to underestimate the worst case performance penalty of a sync.Pools approach.

**Comment by Oppen** [comment]
That's true. It's a bit of a black box and it needs some kind of synchronization. Does Go have thread-local storage? I'm guessing not, but it'd be cool.

**Comment by Oppen** [comment]
Anyway, there may or may not be a gain compared to allocation.

**Comment by Oppen** [comment]
It's contained enough to test several alternatives I think.

**Comment by guymolinari** [comment]
We have so many brilliant minds focusing on this code.  Wish we could find
a way to get the C/C++ libraries running on GPUs.

**Comment by Oppen** [comment]
That would be an interesting project indeed.

**Comment by guymolinari** [comment]
I did a quick, hacked together POC about a year ago.   Roaring read
operations would see a 3 - 5x improvement.

**Comment by Oppen** [comment]
Nice. Is it in GitHub?

**Comment by guymolinari** [comment]
No, but I will get it out there.   I'm overwhelmed and haven't had the time
to take it further.

**Comment by Oppen** [comment]
I think the sooner you publish, the better. That way someone can help.

**Comment by anacrolix** [comment]
How about just exposing Iterator.Init so the caller manages the allocation?

**Comment by lemire** [comment]
@Oppen Can we close this issue?

**Comment by anacrolix** [comment]
Is the PR available in a release or just master? I have some code that uses iterators heavily, I can test the PR fixes it if that's not a certainty yet?

**Comment by lemire** [comment]
@anacrolix It is not yet in a release.

What about this... run tests using our current state and if it helps your problems, then I will issue a release... if not, we will investigate further.

Does that work for you?

**Comment by anacrolix** [comment]
I'm a bit surprised but it didn't fix the problem. In fact it's `roaring.newIntIterator` that contributes the most to the heap allocations. I tried a basic change to expose `IntIterator`, and `IntIterator.Init` for value allocation, but I think Go still decides to escape to the heap due to the complexity of `intIterator.init`. Maybe it's necessary to have a constructor return a value type to avoid this escape.

In my code I've implemented a workaround by caching the iterators, but it's a bit hacky since you can only resume iterators that stopped before where you want to advance from: https://github.com/anacrolix/torrent/commit/df947887fc5d90c2272edc7d02299e92068113c2#diff-449c382592420ab0e94eb0707f863e9fdaeceea483a1f39ebce93c42b57ba04dR149.

**Comment by anacrolix** [comment]
If the caller assumes responsibility for managing the IntIterator, allocations drop to zero: https://github.com/RoaringBitmap/roaring/compare/master...anacrolix:recycleable-int-iterator?expand=1. Here, everytime an iterator is required, `IntIterator.Init` is called with the target `Bitmap`. I don't think it's possible to just return a value `IntIterator` because Go's escape analysis doesn't like `intIterator.init` per my last comment. That path would have lead to being able to return a different type from `Bitmap.Iterator` and retain the existing API in the tests :(.

**Comment by lemire** [comment]
@anacrolix That's a distinct issue. The issue was "Currently, iterators are implemented by instantiating sub-iterators that run over the containers. This approach requires us to create many more temporary objects than necessary." What you refer to is the fact that when you create an iterator, it is going to escape to the heap... which is what is expected. What we fixed was that this iterator, once created, will no longer cause new allocations when you use it.

Specifically, I am saying that you are describing a *distinct* issue.

Now, your issue is certainly valid.

I have created a PR which could meet your need...

https://github.com/RoaringBitmap/roaring/pull/354
